buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/java/AvailableJavaInstallations.kt

```diff
@@ -66,71 +62,73 @@ private
-        javaInstallationForTest = when (javaHomeForTest) {
-            null -> currentJavaInstallation
-            else -> detectJavaInstallation(javaHomeForTest)
-        }
+        javaInstallationForTest = determineJavaInstallation(testJavaHomePropertyName)
+        javaInstallationForCompilation = determineJavaInstallation(compileJavaHomePropertyName)
     }
 
-    fun jdkForCompilation(javaVersion: JavaVersion) = when {
-        javaVersion == currentJavaInstallation.javaVersion -> currentJavaInstallation
-        javaInstallations.containsKey(javaVersion) -> javaInstallations[javaVersion]!!
-        javaVersion <= currentJavaInstallation.javaVersion -> currentJavaInstallation
-        else -> throw IllegalArgumentException("No Java installation found which supports Java version $javaVersion")
+    private
+    fun determineJavaInstallation(propertyName: String): JavaInstallation {
+        val resolvedJavaHome = resolveJavaHomePath(propertyName)
+        when (resolvedJavaHome) {
+            null -> return currentJavaInstallation
+            else -> return detectJavaInstallation(resolvedJavaHome)
+        }
     }
 
-    fun validateForRemoteBuildCacheUsage() {
-        val validationErrors = validateCompilationJdks()
-        if (validationErrors.isNotEmpty()) {
-            val message = formatValidationError(
-                "In order to have cache hits from the remote build cache, your environment needs to be configured accordingly!",
-                validationErrors
-            )
-            throw GradleException(message)
+    fun validateForAllBuilds() {
+        if (project.gradle.startParameter.isBuildCacheEnabled && currentJavaInstallation.vendorAndMajorVersion != oracleJdk8) {
+            System.err.println("To leverage remote cache, you'd better use Oracle JDK 8 to perform this build. It's currently Is currently ${currentJavaInstallation.vendorAndMajorVersion} at ${currentJavaInstallation.javaHome}")
         }
+
+        validate(validateBuildJdks())
+    }
+
+    fun validateForCompilation() {
+        validate(validateCompilationJdks())
     }
 
     fun validateForProductionEnvironment() {
-        val validationErrors = validateCompilationJdks() +
-            mapOf(
-                validationMessage(testJavaHomePropertyName, javaInstallationForTest, oracleJdk8) to (javaInstallationForTest.vendorAndMajorVersion != oracleJdk8)
-            ).filterValues { it }.keys
-        if (validationErrors.isNotEmpty()) {
-            throw GradleException(formatValidationError("JDKs not configured correctly for production build.", validationErrors))
+        validate(validateProductionJdks())
+    }
+
+    private
+    fun validate(errorMessages: Map<String, Boolean>) {
+        val errors = errorMessages.filterValues { it }.keys
+        if (errors.isNotEmpty()) {
+            throw GradleException(formatValidationError("JDKs not configured correctly for production build.", errors))
```

Comment:
`production build` is not correct any more. How about using here `JDKs not configured correctly for build.` and print the production warning in the corresponding error message.